src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs

```diff
@@ -0,0 +1,2408 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Text.Json
+{
+    public ref partial struct Utf8JsonReader
+    {
+        /// <summary>
+        /// Constructs a new <see cref="Utf8JsonReader"/> instance.
+        /// </summary>
+        /// <param name="jsonData">The ReadOnlySequence&lt;byte&gt; containing the UTF-8 encoded JSON text to process.</param>
+        /// <param name="isFinalBlock">True when the input span contains the entire data to process.
+        /// Set to false only if it is known that the input span contains partial data with more data to follow.</param>
+        /// <param name="state">If this is the first call to the ctor, pass in a default state. Otherwise,
+        /// capture the state from the previous instance of the <see cref="Utf8JsonReader"/> and pass that back.</param>
+        /// <remarks>
+        /// Since this type is a ref struct, it is a stack-only type and all the limitations of ref structs apply to it.
+        /// This is the reason why the ctor accepts a <see cref="JsonReaderState"/>.
+        /// </remarks>
+        public Utf8JsonReader(in ReadOnlySequence<byte> jsonData, bool isFinalBlock, JsonReaderState state)
+        {
+            _buffer = jsonData.First.Span;
+
+            _isFinalBlock = isFinalBlock;
+
+            // Note: We do not retain _bytesConsumed or _sequencePosition as they reset with the new input data
+            _allocationFreeContainer = state._allocationFreeContainer;
+            _lineNumber = state._lineNumber;
+            _bytePositionInLine = state._bytePositionInLine;
+            _currentDepth = state._currentDepth;
+            _maxDepth = state._maxDepth == 0 ? JsonReaderState.AllocationFreeMaxDepth : state._maxDepth; // If max depth is not set, revert to the default depth.
+            _inObject = state._inObject;
+            _isNotPrimitive = state._isNotPrimitive;
+            _tokenType = state._tokenType;
+            _previousTokenType = state._previousTokenType;
+            _readerOptions = state._readerOptions;
+            _bitArray = state._bitArray;
+
+            _consumed = 0;
+            _totalConsumed = 0;
+
+            ValueSpan = ReadOnlySpan<byte>.Empty;
+
+            _sequence = jsonData;
+            HasValueSequence = false;
+            ValueSequence = ReadOnlySequence<byte>.Empty;
+
+            if (jsonData.IsSingleSegment)
+            {
+                _nextPosition = default;
+                _currentPosition = default;
+                _isLastSegment = isFinalBlock;
+                _isSingleSegment = true;
+            }
+            else
+            {
+                _nextPosition = jsonData.Start;
+                if (_buffer.Length == 0)
```

Comment:
Why is this check needed?  If it's because the first segment of the sequence might be empty, do we not also need to care about a case where the first segment isn't empty but its start position is pointing to the end of the first segment?  Or is that not possible?

Reply:
> Why is this check needed? If it's because the first segment of the sequence might be empty, 

Yes.

> do we not also need to care about a case where the first segment isn't empty but its start position is pointing to the end of the first segment? Or is that not possible?

We don't need to check for that explicitly since the length check covers that case already. We are getting `_buffer` by calling `ReadOnlySequence.First.Span` and First already does the slicing based on start position. So, if we have a segment with the start position at the end, the span we get will have length 0.
https://github.com/dotnet/corefx/blob/ec7b5ed742b0bb5a233d31f472be082ce57e63ad/src/System.Memory/src/System/Buffers/ReadOnlySequence.Helpers.cs#L112-L117

cc @pakrym